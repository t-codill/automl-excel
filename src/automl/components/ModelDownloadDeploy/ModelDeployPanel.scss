.panel-header {
    padding-left: 40px;
}