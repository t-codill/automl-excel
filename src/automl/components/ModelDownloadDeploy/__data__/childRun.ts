import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
    }
};
