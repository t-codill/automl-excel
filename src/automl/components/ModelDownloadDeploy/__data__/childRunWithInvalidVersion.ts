import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRunWithInvalidVersion: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        dependencies_versions: "{\"test\": \"1.0.0\"}"
    }
};
