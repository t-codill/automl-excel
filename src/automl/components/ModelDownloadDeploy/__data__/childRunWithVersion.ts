import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRunWithVersion: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        dependencies_versions: "{\"azureml-train-automl\": \"1.0.0\"}"
    }
};
