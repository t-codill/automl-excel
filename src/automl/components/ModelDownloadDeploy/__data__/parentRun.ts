import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id",
    properties: {
        DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
    }
};
