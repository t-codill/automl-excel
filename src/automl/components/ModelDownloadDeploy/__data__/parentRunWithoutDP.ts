import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const parentRunWithoutDP: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id",
    properties: {
    }
};
