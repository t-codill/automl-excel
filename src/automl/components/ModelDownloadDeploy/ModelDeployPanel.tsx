import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { DefaultButton, Icon, IPanelProps, Link, MessageBar, MessageBarType, Panel, PanelType, PrimaryButton, ProgressIndicator } from "office-ui-fabric-react";
import * as React from "react";
import { PageNames } from "../../common/PageNames";
import { getCondaFileFromTemplate } from "../../common/utils/deployment/getCondaFileFromTemplate";
import { getScoringFileFromTemplate } from "../../common/utils/deployment/getScoringFileFromTemplate";
import { hasDefaultScoring } from "../../common/utils/deployment/hasDefaultScoring";
import { getModelNameFromRunId } from "../../common/utils/getModelNameFromRunId";
import { getThemeNeutralPrimary } from "../../common/utils/theme/getThemeNeutralPrimary";
import { ArtifactService } from "../../services/ArtifactService";
import { ModelManagementService } from "../../services/ModelManagementService";
import { BaseComponent } from "../Base/BaseComponent";
import { Form } from "../Form/Form";
import { FormTextInput } from "../Form/FormTextInput";
import { Validators } from "../Form/Validators";
import { CondaFileSelector } from "./CondaFileSelector";
import { IDeployModelParams } from "./IDeployModelParams";
import { ScoringScriptSelector } from "./ScoringScriptSelector";

import "./ModelDeployPanel.scss";

export interface IModelDeployPanelProps {
    pageName: PageNames;
    experimentName: string | undefined;
    run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined;
    parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined;
    modelUri: string | undefined;
    scoringUri: string | undefined | null;
    condaUri: string | undefined | null;
    onCancel(): void;
    onModelDeploy(operationId: string): void;
}

export interface IModelDeployPanelState {
    deploying: boolean;
    modelId: string | undefined;
    primaryMetric: string | undefined;
    autoGenerateScoringFile: boolean;
    autoGenerateCondaFile: boolean;
    errorMessage: string | undefined;
}

export class ModelDeployPanel extends BaseComponent<IModelDeployPanelProps, IModelDeployPanelState, {
    modelManagementService: ModelManagementService;
    artifactService: ArtifactService;
}> {
    protected serviceConstructors = {
        modelManagementService: ModelManagementService,
        artifactService: ArtifactService
    };
    protected getData = undefined;

    constructor(props: IModelDeployPanelProps) {
        super(props);
        this.state = {
            deploying: false,
            modelId: undefined,
            primaryMetric: this.props.parentRun && this.props.parentRun.properties && this.props.parentRun.properties.primary_metric,
            autoGenerateScoringFile: true,
            autoGenerateCondaFile: true,
            errorMessage: undefined
        };
    }

    public readonly render = (): React.ReactNode => {
        return <>
            <Panel
                isOpen={true}
                type={PanelType.medium}
                onDismiss={this.props.onCancel}
                isHiddenOnDismiss={true}
                onRenderHeader={this.renderHeader}
                closeButtonAriaLabel="Close">
                <Form onSubmit={this.deployModel}>
                    {this.state.errorMessage ? <MessageBar messageBarType={MessageBarType.error}>{this.state.errorMessage}</MessageBar> : <></>}
                    <FormTextInput<IDeployModelParams, "name">
                        field="name"
                        required={true}
                        label="Deployment name"
                        placeholder="Deployment name"
                        ariaLabel="Please enter deployment name here"
                        validators={[
                            Validators.required("Model name is required"),
                            Validators.regex(/^[a-z][a-z0-9-]{1,15}$/,
                                "Model name should start with a lower case letter, be between 2 and 16 character, and only include letters (a-z), numbers (0-9) and '-'")]}
                    />

                    <FormTextInput<IDeployModelParams, "description">
                        field="description"
                        required={false}
                        multiline={true}
                        label="Deployment description"
                        placeholder="Deployment description"
                        ariaLabel="Please enter deployment description here"
                    />

                    <p>Target: Only Azure Container Instance(ACI) is supported</p>
                    <ScoringScriptSelector onToggle={this.handleScoringToggle} hasDefaultOption={hasDefaultScoring(this.props.parentRun, this.props.scoringUri)} />
                    <CondaFileSelector onToggle={this.handleCondaToggle} />
                    {this.state.deploying ?
                        <ProgressIndicator description="Deploying Model..." /> : undefined}
                    <div className="form-footer">
                        <DefaultButton text="Cancel" onClick={this.props.onCancel} />
                        <PrimaryButton text="Deploy" disabled={this.state.deploying} type="submit" />
                    </div>
                </Form>
            </Panel>
        </>;
    }

    private readonly registerModel = async (run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto,
        modelUri: string, experimentName: string) => {
        const asset = await this.services.modelManagementService.createAsset(
            experimentName,
            `${run.runId}_Model`,
            modelUri
        );
        this.logUserAction("RegisterModel",
            { modelId: this.state.modelId, experimentName: this.props.experimentName, runId: run.runId });
        if (!asset || !asset.id || !run.runId) {
            return;
        }
        const model = await this.services.modelManagementService.registerModel(
            getModelNameFromRunId(run.runId),
            `${run.runId}_Model`,
            asset.id,
            "application/json",
            run.runId,
            experimentName
        );
        if (!model) {
            this.setState({ errorMessage: "Register Model Failed. Please check logs" });
            return;
        }
        const modelId = model.id;
        if (!modelId) {
            this.setState({ errorMessage: "Model Id is invalid. Please check logs" });
            return;
        }
        this.setState({ modelId });
    }

    private readonly renderHeader = (props?: IPanelProps) => {
        const color = getThemeNeutralPrimary(props);
        return (
            this.props.pageName === PageNames.ParentRun ?
                <div className="panel-header" style={{
                    color
                }}>
                    <h2>Deploy Best Model</h2>
                    <p>Best model based on {this.state.primaryMetric}</p>
                </div> :
                <div className="panel-header" style={{
                    color
                }}>
                    <h2>Deploy Model</h2>
                    <p>Deploy the model to a web service which can then be used to predict on new data. &nbsp;
                        <Link target="_blank" href="https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-create-portal-experiments#deploy-model">
                            Learn more <Icon iconName="NavigationExternalInline" />
                        </Link>
                    </p>
                </div>
        );
    }
    private readonly handleCondaToggle = () => {
        this.setState((prev) => {
            return {
                autoGenerateCondaFile: !prev.autoGenerateCondaFile,
            };
        });
    }

    private readonly handleScoringToggle = () => {
        this.setState((prev) => {
            return {
                autoGenerateScoringFile: !prev.autoGenerateScoringFile,
            };
        });
    }

    private readonly readOrUploadCondaFile = async (runId: string, data: IDeployModelParams): Promise<string | undefined> => {
        const condaFilePath = "ModelDeploy/condaEnv.yml";
        if (this.state.autoGenerateCondaFile
            && this.props.run
            && this.props.run.properties
            && this.props.run.properties.conda_env_data_location) {
            return this.props.run.properties.conda_env_data_location;
        }
        const condaFileContent = this.state.autoGenerateCondaFile ? getCondaFileFromTemplate(this.props.run) : data.condaFile;
        if (!condaFileContent) {
            return undefined;
        }
        const condaArtifactId = await this.services.artifactService.uploadArtifact(`dcid.${runId}`, condaFilePath, condaFileContent);
        if (!condaArtifactId) {
            return undefined;
        }
        return this.getArtifactUrl(condaArtifactId);
    }

    private readonly readOrUploadScoringFile = async (runId: string, data: IDeployModelParams): Promise<string | undefined> => {
        const scoringFilePath = "ModelDeploy/scoring.py";
        if (this.state.autoGenerateScoringFile
            && this.props.run
            && this.props.run.properties
            && this.props.run.properties.scoring_data_location) {
            return this.props.run.properties.scoring_data_location;
        }
        const scoringFileContent = this.state.autoGenerateScoringFile ? getScoringFileFromTemplate(this.props.parentRun, runId) : data.scoringFile;
        if (!scoringFileContent) {
            return undefined;
        }
        const scoringArtifactId = await this.services.artifactService.uploadArtifact(`dcid.${runId}`, scoringFilePath, scoringFileContent);
        if (!scoringArtifactId) {
            return undefined;
        }
        return this.getArtifactUrl(scoringArtifactId);
    }

    private readonly getArtifactUrl = (artifactId: string): string => {
        return `aml://artifact/${artifactId}`;
    }

    // tslint:disable-next-line: cyclomatic-complexity
    private readonly deployModel = async (data: IDeployModelParams) => {
        const runId = this.props.run && this.props.run.runId;
        if (!this.props.run || !this.props.run.runId || !runId || !this.props.experimentName || !this.props.modelUri) {
            this.setState({ deploying: false });
            return;
        }
        this.setState({ deploying: true });
        this.logUserAction("DeployModel",
            { deployName: data.name, autoGenerateCondaFile: this.state.autoGenerateCondaFile, autoGenerateScoringFile: this.state.autoGenerateScoringFile });
        await this.registerModel(this.props.run, this.props.modelUri, this.props.experimentName);
        if (!this.state.modelId) {
            this.setState({ deploying: false, errorMessage: "Register Model Failed" });
            return;
        }

        const condaArtifactLocation = await this.readOrUploadCondaFile(runId, data);
        const scoringArtifactLocation = await this.readOrUploadScoringFile(runId, data);
        if (!condaArtifactLocation || !scoringArtifactLocation) {
            this.setState({ deploying: false, errorMessage: "Upload File to Artifact Failed" });
            return;
        }

        const scoringFileName = scoringArtifactLocation.substring(scoringArtifactLocation.lastIndexOf("/") + 1);
        const response = await this.services.modelManagementService.createDeployment(data.name, data.description, runId,
            this.state.modelId,
            condaArtifactLocation,
            scoringFileName,
            scoringArtifactLocation);
        if (response && response.status === 409) {
            this.setState({
                deploying: false,
                errorMessage: `Current workspace already contains a deployment with name ${data.name}`
            });
            return;
        }
        if (!response || response.status !== 202 || !response.operationId) {
            this.setState({
                deploying: false,
                errorMessage: "Create Deployment Failed"
            });
            return;
        }
        this.props.onModelDeploy(response.operationId);
    }
}
