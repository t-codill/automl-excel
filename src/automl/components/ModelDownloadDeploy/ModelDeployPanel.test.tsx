import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { shallow, ShallowWrapper } from "enzyme";
import { Link, Panel } from "office-ui-fabric-react";
import * as React from "react";
import { getLogCustomEventSpy, testContext } from "../../common/context/__data__/testContext";
import { PageNames } from "../../common/PageNames";
import { blob2string } from "../../common/utils/blob2string";
import * as saveAsModule from "../../common/utils/saveAs";
import { defaultSdkVersion } from "../../services/JasmineServiceJsonDefinition";
import { ModelManagementService } from "../../services/ModelManagementService";
import { RunHistoryService } from "../../services/RunHistoryService";
import { BaseComponent } from "../Base/BaseComponent";
import { IModelDeployPanelProps, IModelDeployPanelState, ModelDeployPanel } from "./ModelDeployPanel";

jest.mock("../../services/ModelManagementService");
jest.mock("../../services/RunHistoryService");

describe("Model Deploy Panel", () => {
    const parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
        runId: "test-run-id",
        properties: {
            DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
        }
    };
    const childRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
        runId: "test-run-id_1",
        properties: {
            DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
        }
    };
    let asCreateAsset: jest.SpyInstance<ReturnType<ModelManagementService["createAsset"]>>;
    let asRegisterModel: jest.SpyInstance<ReturnType<ModelManagementService["registerModel"]>>;
    let asUpdateTag: jest.SpyInstance<ReturnType<RunHistoryService["updateTag"]>>;
    const mockOnCancel = jest.fn();
    const mockOnModelRegister = jest.fn();

    describe("Missing Data", () => {
        beforeEach(() => {
            asCreateAsset = jest.spyOn(ModelManagementService.prototype, "createAsset");
            asRegisterModel = jest.spyOn(ModelManagementService.prototype, "registerModel");
            asUpdateTag = jest.spyOn(RunHistoryService.prototype, "updateTag");
        });

        it("No Run", async () => {
            const invalidProps: IModelDeployPanelProps = {
                experimentName: "test-experiment-name",
                run: undefined,
                parentRun: undefined,
                modelUri: undefined,
                modelId: undefined,
                onCancel: mockOnCancel,
                onModelRegister: mockOnModelRegister
            };
            const tree = shallow<ModelDeployPanel>(
                <ModelDeployPanel {...invalidProps} />);
            tree.find(Link)
                .filterWhere((l) => l.key() === "register")
                .simulate("click");
            await Promise.resolve();
            expect(asCreateAsset)
                .toBeCalledTimes(0);
        });

        it("No ModelUri ", async () => {
            const invalidProps: IModelDeployPanelProps = {
                experimentName: "test-experiment-name",
                run: childRun,
                parentRun,
                modelUri: undefined,
                modelId: undefined,
                onCancel: mockOnCancel,
                onModelRegister: mockOnModelRegister
            };
            const tree = shallow<ModelDeployPanel>(
                <ModelDeployPanel {...invalidProps} />);
            tree.find(Link)
                .filterWhere((l) => l.key() === "register")
                .simulate("click");
            await Promise.resolve();
            expect(asCreateAsset)
                .toBeCalledTimes(0);

        });

        it("No ExperimentName ", async () => {
            const invalidProps: IModelDeployPanelProps = {
                experimentName: undefined,
                run: childRun,
                parentRun,
                modelUri: "test-uri",
                modelId: undefined,
                onCancel: mockOnCancel,
                onModelRegister: mockOnModelRegister
            };
            const tree = shallow<ModelDeployPanel>(
                <ModelDeployPanel {...invalidProps} />);
            tree.find(Link)
                .filterWhere((l) => l.key() === "register")
                .simulate("click");
            await Promise.resolve();
            expect(asCreateAsset)
                .toBeCalledTimes(0);

        });
    });

    describe("Model is not generated", () => {
        const props: IModelDeployPanelProps = {
            experimentName: "test-experiment-name",
            run: childRun,
            parentRun,
            modelUri: "https://demo.azure.com/model_123",
            modelId: undefined,
            onCancel: mockOnCancel,
            onModelRegister: mockOnModelRegister
        };
        let tree: ShallowWrapper<IModelDeployPanelProps, IModelDeployPanelState>;
        let register: ((...[]) => void);
        beforeEach(async () => {
            asCreateAsset = jest.spyOn(ModelManagementService.prototype, "createAsset");
            asRegisterModel = jest.spyOn(ModelManagementService.prototype, "registerModel");
            asUpdateTag = jest.spyOn(RunHistoryService.prototype, "updateTag");
            tree = shallow(
                <ModelDeployPanel {...props} />
            );
            await Promise.resolve();
            register = tree.find(Link)
                .filterWhere((l) => l.key() === "register")
                .prop("onClick") || jest.fn();
        });
        it("Should render register model", () => {
            expect(tree)
                .toMatchSnapshot();
        });
        it("Should render header", () => {
            const onRenderHeader = tree.find(Panel)
                .prop("onRenderHeader");
            let header: React.ReactNode | undefined;
            if (onRenderHeader) {
                header = onRenderHeader();
            }
            expect(header)
                .toMatchSnapshot();
        });
        it("Should render header for parent", async () => {
            jest.spyOn(BaseComponent.prototype.context, "pageName", "get")
                .mockReturnValue(PageNames.ParentRun);
            await Promise.resolve();
            const onRenderHeader = tree.find(Panel)
                .prop("onRenderHeader");
            let header: React.ReactNode | undefined;
            if (onRenderHeader) {
                header = onRenderHeader();
            }
            expect(shallow(<div>
                {header}
            </div>)
                .find("h2")
                .text())
                .toBe("Deploy Best Model");
        });
        describe("register model", () => {
            it("should call create asset", async () => {
                asCreateAsset.mockReturnValueOnce(Promise.resolve(undefined));
                const logSpy = getLogCustomEventSpy();
                register();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_RegisterModel_UserAction",
                        testContext,
                        { component: "RegisterModel", experimentName: "test-experiment-name", modelId: undefined, pageName: "", runId: "test-run-id_1" }
                    );
                expect(asCreateAsset)
                    .toBeCalledWith(
                        "test-experiment-name",
                        "test-run-id_1_Model",
                        "https://demo.azure.com/model_123"
                    );
                expect(asRegisterModel)
                    .toBeCalledTimes(0);
                expect(asUpdateTag)
                    .toBeCalledTimes(0);
                expect(mockOnModelRegister)
                    .toBeCalledTimes(0);

            });
            it("should not call proceeded if asset id missing", async () => {
                asCreateAsset.mockReturnValueOnce(Promise.resolve({
                    name: "asset-name"
                }));
                const logSpy = getLogCustomEventSpy();
                register();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_RegisterModel_UserAction",
                        testContext,
                        { component: "RegisterModel", experimentName: "test-experiment-name", modelId: undefined, pageName: "", runId: "test-run-id_1" }
                    );
                expect(asCreateAsset)
                    .toBeCalledWith(
                        "test-experiment-name",
                        "test-run-id_1_Model",
                        "https://demo.azure.com/model_123"
                    );
                expect(asRegisterModel)
                    .toBeCalledTimes(0);
                expect(mockOnModelRegister)
                    .toBeCalledTimes(0);
                expect(asUpdateTag)
                    .toBeCalledTimes(0);

            });
            it("should call register model", async () => {
                asRegisterModel.mockReturnValueOnce(Promise.resolve(undefined));
                const logSpy = getLogCustomEventSpy();
                register();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_RegisterModel_UserAction",
                        testContext,
                        { component: "RegisterModel", experimentName: "test-experiment-name", modelId: undefined, pageName: "", runId: "test-run-id_1" }
                    );
                expect(asRegisterModel)
                    .toBeCalledWith(
                        "test-experiment-name",
                        "test-run-id_1_Model",
                        "asset-id",
                        "application/json",
                        "test-run-id_1",
                        "test-experiment-name"
                    );
                expect(asUpdateTag)
                    .toBeCalledTimes(0);
                expect(mockOnModelRegister)
                    .toBeCalledTimes(0);

            });
            it("should call update tag", async () => {
                asUpdateTag.mockReturnValueOnce(Promise.resolve(undefined));
                const logSpy = getLogCustomEventSpy();
                register();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_RegisterModel_UserAction",
                        testContext,
                        { component: "RegisterModel", experimentName: "test-experiment-name", modelId: undefined, pageName: "", runId: "test-run-id_1" }
                    );
                expect(asUpdateTag)
                    .toBeCalledWith(
                        {
                            runId: "test-run-id_1",
                            properties: {
                                DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
                            }
                        },
                        "test-experiment-name",
                        "model_id",
                        "model-name"
                    );
                expect(mockOnModelRegister)
                    .toBeCalledTimes(0);
                expect(tree.state("registering"))
                    .toBe(false);

            });

            it("should successes", async () => {
                const logSpy = getLogCustomEventSpy();
                register();
                await Promise.resolve();
                await Promise.resolve();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_RegisterModel_UserAction",
                        testContext,
                        { component: "RegisterModel", experimentName: "test-experiment-name", modelId: undefined, pageName: "", runId: "test-run-id_1" }
                    );
                expect(mockOnModelRegister)
                    .toBeCalledTimes(1);
                expect(tree.state())
                    .toEqual({
                        registering: false,
                        modelId: "model-name"
                    });

            });
        });
    });
    describe("Model is generated", () => {
        const props: IModelDeployPanelProps = {
            experimentName: "test-experiment-name",
            run: childRun,
            parentRun,
            modelUri: "https://demo.azure.com/model_123",
            modelId: "test-model-id",
            onCancel: mockOnCancel,
            onModelRegister: mockOnModelRegister
        };
        let downloadScoringFile: ((...[]) => void);
        let getCondaEnvFile: ((...[]) => void);
        let tree: ShallowWrapper<IModelDeployPanelProps, IModelDeployPanelState>;
        beforeEach(async () => {
            asCreateAsset = jest.spyOn(ModelManagementService.prototype, "createAsset");
            asRegisterModel = jest.spyOn(ModelManagementService.prototype, "registerModel");
            asUpdateTag = jest.spyOn(RunHistoryService.prototype, "updateTag");
            tree = shallow(
                <ModelDeployPanel {...props} />
            );
            await Promise.resolve();
            const links = tree.find(Link);
            const scoring = links
                .filterWhere((l) => l.key() === "scoring")
                .prop("onClick");
            const environment = links
                .filterWhere((l) => l.key() === "environment")
                .prop("onClick");
            downloadScoringFile = scoring || (() => { return; });
            getCondaEnvFile = environment || (() => { return; });
        });
        describe("Validate Render", () => {
            it("should render correctly", () => {
                expect(tree)
                    .toMatchSnapshot();
            });
        });
        it("should show spinning during registering", () => {
            tree.setState({ registering: true });
            expect(tree)
                .toMatchSnapshot();
            tree.setState({ registering: false });
        });

        describe("scoring file", () => {
            it("should download", async () => {
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                const logSpy = getLogCustomEventSpy();
                downloadScoringFile();
                expect(logSpy)
                    .toBeCalledWith(
                        "_DownloadScoringFile_UserAction",
                        testContext,
                        { component: "DownloadScoringFile", experimentName: "test-experiment-name", modelId: "test-model-id", pageName: "", runId: "test-run-id_1" }
                    );

                expect(asSaveAs)
                    .toBeCalledWith(expect.anything(), "scoring.py");
            });

            it("should not downloading if model id missing", async () => {
                tree.setState({ modelId: undefined });
                await Promise.resolve();
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                downloadScoringFile();

                expect(asSaveAs)
                    .toBeCalledTimes(0);
            });

            it("should not downloading if column is missing", async () => {
                tree.setProps({
                    parentRun: {
                        runId: "test-run-id",
                        properties: {}
                    }
                });
                await Promise.resolve();
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                const logSpy = getLogCustomEventSpy();
                downloadScoringFile();
                expect(logSpy)
                    .toBeCalledWith(
                        "_DownloadScoringFile_UserAction",
                        testContext,
                        { component: "DownloadScoringFile", experimentName: "test-experiment-name", modelId: "test-model-id", pageName: "", runId: "test-run-id_1" }
                    );

                expect(asSaveAs)
                    .toBeCalledTimes(0);
            });
        });

        describe("conda env", () => {
            it("should download", () => {
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                const logSpy = getLogCustomEventSpy();
                getCondaEnvFile();
                expect(logSpy)
                    .toBeCalledWith(
                        "_DownloadCondaFile_UserAction",
                        testContext,
                        { component: "DownloadCondaFile", experimentName: "test-experiment-name", pageName: "", runId: "test-run-id_1" }
                    );

                expect(asSaveAs)
                    .toBeCalledWith(expect.anything(), "condaEnv.yml");
            });

            it("should use sdk version from properties", async () => {
                tree.setProps({
                    run: {
                        runId: "test-run-id",
                        properties: {
                            DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}",
                            dependencies_versions: JSON.stringify({
                                "azureml-train-automl": "testSdkVersion"
                            })
                        }
                    }
                });
                await Promise.resolve();
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                const logSpy = getLogCustomEventSpy();
                getCondaEnvFile();
                await Promise.resolve();
                await Promise.resolve();
                expect(logSpy)
                    .toBeCalledWith(
                        "_DownloadCondaFile_UserAction",
                        testContext,
                        { component: "DownloadCondaFile", experimentName: "test-experiment-name", pageName: "", runId: "test-run-id" }
                    );

                expect(asSaveAs)
                    .toBeCalledTimes(1);
                const content = await blob2string(asSaveAs.mock.calls[0][0]);
                expect(content)
                    .toMatch(/azureml-sdk\[automl\]==testSdkVersion/);
            });
            it("should use default version if property is invalid", async (done) => {
                tree.setProps({
                    run: {
                        runId: "test-run-id",
                        properties: {
                            DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}",
                            dependencies_versions: "invalid json string"
                        }
                    }
                });
                await Promise.resolve();
                const asSaveAs = jest.spyOn(saveAsModule, "saveAs");
                const logSpy = getLogCustomEventSpy();
                getCondaEnvFile();
                expect(logSpy)
                    .toBeCalledWith(
                        "_DownloadCondaFile_UserAction",
                        testContext,
                        { component: "DownloadCondaFile", experimentName: "test-experiment-name", pageName: "", runId: "test-run-id" }
                    );

                expect(asSaveAs)
                    .toBeCalledTimes(1);
                expect(await blob2string(asSaveAs.mock.calls[0][0]))
                    .toMatch(new RegExp(`azureml-sdk\\[automl\\]==${defaultSdkVersion}`));
                done();
            });
        });
    });
});
