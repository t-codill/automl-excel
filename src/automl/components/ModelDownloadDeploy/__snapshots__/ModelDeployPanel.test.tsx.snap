// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model Deploy Panel Run with dependency version should toggle conda file 1`] = `
<Fragment>
  <StyledPanelBase
    closeButtonAriaLabel="Close"
    isHiddenOnDismiss={true}
    isOpen={true}
    onDismiss={[MockFunction]}
    onRenderHeader={[Function]}
    type={3}
  >
    <Form
      onSubmit={[Function]}
    >
      <FormTextInput
        ariaLabel="Please enter deployment name here"
        field="name"
        label="Deployment name"
        placeholder="Deployment name"
        required={true}
        validators={
          Array [
            [Function],
            [Function],
          ]
        }
      />
      <FormTextInput
        ariaLabel="Please enter deployment description here"
        field="description"
        label="Deployment description"
        multiline={true}
        placeholder="Deployment description"
        required={false}
      />
      <p>
        Target: Only Azure Container Instance(ACI) is supported
      </p>
      <ScoringScriptSelector
        onToggle={[Function]}
      />
      <CondaFileSelector
        onToggle={[Function]}
      />
      <div
        className="form-footer"
      >
        <CustomizedDefaultButton
          onClick={[MockFunction]}
          text="Cancel"
        />
        <CustomizedPrimaryButton
          disabled={false}
          text="Deploy"
          type="submit"
        />
      </div>
    </Form>
  </StyledPanelBase>
</Fragment>
`;

exports[`Model Deploy Panel Run with dependency version should toggle scoring file 1`] = `
<Fragment>
  <StyledPanelBase
    closeButtonAriaLabel="Close"
    isHiddenOnDismiss={true}
    isOpen={true}
    onDismiss={[MockFunction]}
    onRenderHeader={[Function]}
    type={3}
  >
    <Form
      onSubmit={[Function]}
    >
      <FormTextInput
        ariaLabel="Please enter deployment name here"
        field="name"
        label="Deployment name"
        placeholder="Deployment name"
        required={true}
        validators={
          Array [
            [Function],
            [Function],
          ]
        }
      />
      <FormTextInput
        ariaLabel="Please enter deployment description here"
        field="description"
        label="Deployment description"
        multiline={true}
        placeholder="Deployment description"
        required={false}
      />
      <p>
        Target: Only Azure Container Instance(ACI) is supported
      </p>
      <ScoringScriptSelector
        onToggle={[Function]}
      />
      <CondaFileSelector
        onToggle={[Function]}
      />
      <div
        className="form-footer"
      >
        <CustomizedDefaultButton
          onClick={[MockFunction]}
          text="Cancel"
        />
        <CustomizedPrimaryButton
          disabled={false}
          text="Deploy"
          type="submit"
        />
      </div>
    </Form>
  </StyledPanelBase>
</Fragment>
`;

exports[`Model Deploy Panel Run without dataprep Should render header 1`] = `
<div
  className="panel-header"
  style={
    Object {
      "color": undefined,
    }
  }
>
  <h2>
    Deploy Best Model
  </h2>
  <p>
    Best model based on 
  </p>
</div>
`;
